### PR TITLE
SetImplementation

### PR DESCRIPTION
**Pull request by oleksandr-hk**
collections-task homework(Humeniuk Oleksandr)

### CODE REVIEW

File: src/main/java/co/inventorsoft/academy/collections/model/Range.java [hunk 1]
```diff
@@ -1,60 +1,263 @@
...
 
     public <T1> T1[] toArray(T1[] a) {
         return null;
     }
 
     public boolean add(T t) {
-        return false;
+
+        //changing size if not enough
+        if (size >= 14) {
```
Comment (george-stotle) [inline src/main/java/co/inventorsoft/academy/collections/model/Range.java]: What magic number is **14**?

File: src/main/java/co/inventorsoft/academy/collections/model/Range.java [hunk 2]
```diff
@@ -1,60 +1,263 @@
...
-        return false;
+        for (Object elem : c) {
+            if (!this.contains(elem)) {
+                return false;
+            }
+        }
+        return true;
     }
 
     public boolean addAll(Collection<? extends T> c) {
```
Comment (george-stotle) [inline src/main/java/co/inventorsoft/academy/collections/model/Range.java:176]: Handle when you have duplication.

File: src/main/java/co/inventorsoft/academy/collections/model/Range.java [hunk 3]
```diff
@@ -1,60 +1,263 @@
...
+        if (bucket == null) {
+            if (size==0){
+                buckets[index] = newNode;
+                size++;
+                current=buckets[index];
+                start=buckets[index];
+                return true;
+
+            }else {
+            buckets[index] = newNode;
```
Comment (george-stotle) [inline src/main/java/co/inventorsoft/academy/collections/model/Range.java:126]: Check if you don't have duplication.

File: src/main/java/co/inventorsoft/academy/collections/model/Range.java [hunk 4]
```diff
@@ -1,60 +1,263 @@
...
     public boolean addAll(Collection<? extends T> c) {
-        return false;
+        for (T elem : c) {
+            this.add(elem);
+        }
+        return true;
+
     }
 
     public boolean retainAll(Collection<?> c) {
```
Comment (george-stotle) [inline src/main/java/co/inventorsoft/academy/collections/model/Range.java:184]: Review the logic of the retainAll method.

File: src/main/java/co/inventorsoft/academy/collections/model/Range.java [hunk 5]
```diff
@@ -1,60 +1,263 @@
...
+            }
+        }
         return false;
     }
 
     public boolean removeAll(Collection<?> c) {
+        for (Object elem : c) {
+            this.remove(elem);
+        }
         return false;
```
Comment (george-stotle) [inline src/main/java/co/inventorsoft/academy/collections/model/Range.java]: Why do you return always false? You have to return true if at least one element was deleted, otherwise false.

File: src/main/java/co/inventorsoft/academy/collections/model/Range.java [hunk 6]
```diff
@@ -1,60 +1,263 @@
...
     }
 
     public void clear() {
+        Node[] newBuckets = new Node[size];
```
Comment (george-stotle) [inline src/main/java/co/inventorsoft/academy/collections/model/Range.java]: Some Garbage Collectors use number of point to element, so go through all elements and set **data** and **next** to null.

File: src/main/java/co/inventorsoft/academy/collections/model/Range.java [hunk 7]
```diff
@@ -1,60 +1,263 @@
...
+        } else {
+            newRange = new Range<Float>();
+            for (float i = start; i <= end; i+=0.1) {
+                newRange.add(i);
+            }
+        }
+        return newRange;
+    }
+
+    static Range<Double> of(double e1, double e2) {
```
Comment (george-stotle) [inline src/main/java/co/inventorsoft/academy/collections/model/Range.java]: What will happen if pass 3.4234?

File: src/main/java/co/inventorsoft/academy/collections/model/Range.java [hunk 8]
```diff
@@ -1,60 +1,263 @@
...
+        } else {
+            newRange = new Range<Double>();
+            for (double i = start; i <= end; i+=0.1) {
+                newRange.add(i);
+            }
+        }
+        return newRange;
+    }
+
+    static Range<Character> of(Character e1, Character e2,Function<Character,Character> function) {
```
Comment (george-stotle) [inline src/main/java/co/inventorsoft/academy/collections/model/Range.java]: 1. Think about **custom** realization of **Number**. 
2. Try to avoid duplication of code. Each overloaded method **of** contains similar code.
3. Catch a case when e2 < e1.

File: src/main/java/co/inventorsoft/academy/collections/model/Range.java [hunk 9]
```diff
@@ -1,60 +1,263 @@
...
+            private int currentIndex = 0;
+
+            @Override
+            public boolean hasNext() {
+                return currentIndex < size;
+            }
+
+            @Override
+            public T next() {
+                Object tmp = start;
```
Comment (george-stotle) [inline src/main/java/co/inventorsoft/academy/collections/model/Range.java:62]: Don't you have to go through all buckets?

File: src/main/java/co/inventorsoft/academy/collections/model/Main.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
+package co.inventorsoft.academy.collections.model;
```
Comment (george-stotle) [inline src/main/java/co/inventorsoft/academy/collections/model/Main.java]: Remove this file

File: src/main/java/co/inventorsoft/academy/collections/model/Range.java (hunk 9 above)
Comment (oleksandr-hk) [reply src/main/java/co/inventorsoft/academy/collections/model/Range.java:62]: What do you mean?

Comment (oleksandr-hk) [reply src/main/java/co/inventorsoft/academy/collections/model/Range.java:62]: Could you rephrase your question?

File: src/main/java/co/inventorsoft/academy/collections/model/Range.java (hunk 2 above)
Comment (oleksandr-hk) [reply src/main/java/co/inventorsoft/academy/collections/model/Range.java:176]: In containsAll method?

File: src/main/java/co/inventorsoft/academy/collections/model/Range.java [hunk 11]
```diff
@@ -1,60 +1,263 @@
...
+            }
+        }
+        return array;
     }
 
     public <T1> T1[] toArray(T1[] a) {
         return null;
     }
 
     public boolean add(T t) {
```
Comment (george-stotle) [inline src/main/java/co/inventorsoft/academy/collections/model/Range.java:99]: Do you check if you have duplication elements?

File: src/main/java/co/inventorsoft/academy/collections/model/Range.java (hunk 2 above)
Comment (george-stotle) [reply src/main/java/co/inventorsoft/academy/collections/model/Range.java:176]: This method returns boolean, that means true or false, but you always return true. Review the logic of the method.

File: src/main/java/co/inventorsoft/academy/collections/model/Range.java (hunk 9 above)
Comment (george-stotle) [reply src/main/java/co/inventorsoft/academy/collections/model/Range.java:62]: So, you have **private Node[] buckets;** do you go through all of them?

Comment (oleksandr-hk) [reply src/main/java/co/inventorsoft/academy/collections/model/Range.java:62]: Depends on situation.If I understood you - yes.

File: src/main/java/co/inventorsoft/academy/collections/model/Range.java (hunk 11 above)
Comment (oleksandr-hk) [reply src/main/java/co/inventorsoft/academy/collections/model/Range.java:99]: No, because Set can't save duplicate elements.Or I didn't understand you.

Comment (george-stotle) [reply src/main/java/co/inventorsoft/academy/collections/model/Range.java:99]: Yeah, right it can't. So if that mustn't contain duplication, do you check if I don't pass the same object twice?
